tests/sentry/api/test_event_search.py

```diff
@@ -849,15 +847,79 @@ def test_multiple_statements(self):
                         left_term=self.term3,
                         operator='AND',
                         right_term=SearchBoolean(
-                            left_term=term4,
+                            left_term=self.term4,
                             operator='AND',
-                            right_term=term5
+                            right_term=self.term5
                         )
                     )
                 )
             )
         )]
 
+    def test_grouping_simple(self):
+        result = parse_search_query(
+            '(user.email:foo@example.com OR user.email:bar@example.com)'
+        )
+        assert result == [SearchBoolean(left_term=self.term1, operator="OR", right_term=self.term2)]
+        result = parse_search_query(
+            '(user.email:foo@example.com OR user.email:bar@example.com) AND user.email:foobar@example.com'
+        )
+        assert result == [SearchBoolean(
+            left_term=SearchBoolean(
+                left_term=self.term1,
+                operator='OR',
+                right_term=self.term2),
+            operator='AND',
+            right_term=self.term3)]
+
+        result = parse_search_query(
+            'user.email:foo@example.com AND (user.email:bar@example.com OR user.email:foobar@example.com)'
+        )
+        assert result == [SearchBoolean(
+            left_term=self.term1,
+            operator='AND',
+            right_term=SearchBoolean(
+                left_term=self.term2,
+                operator='OR',
+                right_term=self.term3))]
+
+    def test_nested_grouping(self):
+        result = parse_search_query(
+            '(user.email:foo@example.com OR (user.email:bar@example.com OR user.email:foobar@example.com))'
+        )
+        assert result == [SearchBoolean(
+            left_term=self.term1,
+            operator="OR",
+            right_term=SearchBoolean(
+                left_term=self.term2,
+                operator="OR",
+                right_term=self.term3))
+        ]
+        result = parse_search_query(
+            '(user.email:foo@example.com OR (user.email:bar@example.com OR (user.email:foobar@example.com AND user.email:hello@example.com OR user.email:hi@example.com)))'
+        )
+        assert result == [SearchBoolean(
+            left_term=self.term1,
+            operator="OR",
+            right_term=SearchBoolean(
+                left_term=self.term2,
+                operator="OR",
+                right_term=SearchBoolean(
+                    left_term=SearchBoolean(
+                        left_term=self.term3,
+                        operator="AND",
+                        right_term=self.term4
+                    ),
+                    operator="OR",
+                    right_term=self.term5,
+                )
+            )
+        )]
+
+    def test_malformed_groups(self):
```

Comment:
Unfortunately I'm not sure what to do about poorly formatted groups. It causes the parser to assume that everything is a message/filter and throws everything off.

Reply:
Nvm. Just going to remove parenthesis from being valid raw strings. If someone wants to use it in raw they can use a quoted raw. Let me know if people have objections to this.

Reply:
Allowing `(` and `)` in bare strings makes it really hard to discern where groups are. I think only allowing them in quoted strings is a good decision.

Reply:
I don't think we can change the expected format of raw strings, we'll break people's searches for sure. I can see 1k+ recent searches that use `(` or `)` in their raw search. People expect to be able to just paste their errors and have it work

Reply:
@wedamija I don't believe there is any way at all to make that work, unless we use a different character for `(` `)` or escape it in some way... which I think would be more confusing. I think those with parenthesis in raw search will just have to change their search to a quoted raw search. We could create a migration to do that. But allowing `(` or `)` into unquoted raw searches creates ambiguity that I don't think there's a way to recover from.

Reply:
Could we require whitespace around the `(` and `)` to help disambiguate the grammar? It makes grouping expressions a bit more clunky but might allow us to not break existing searches.

Reply:
@lauryndbrown If we want to break existing searches it's something that we need to bring up with product - we can't just make an engineering decision here.

@markstory's suggestion could work, although might still not be perfect. Another option could be to allow the users to enable/disable boolean search, but that kind of sucks since we need to maintain multiple code branches.

Are we able to look at patterns here? Like `(` can really only be before a term, like `(tag:1`, or `(hello`. So if we check that all open parens start with either whitespace or beginning of line, and all `)` have whitespace or end of line after them, would that work?

Does it also make sense to say that a group can't being in the middle of a term? So if we had `hello i ( am a parentheses` could we just ignore it since a term has already started? I guess that might be tricky.

One other thought is that if there's no closing parentheses in the right place, maybe we can infer that it's not part of a group?

Reply:
For now I will go with Mark's suggestion.

Reply:
As per conversations with Sara and Dan, we'll go ahead and accept that we will break some existing user's queries. and simply do the more intuitive way with no requirement on spacing.